lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -259,6 +259,19 @@ static bool recursiveType(const StructType *ST, const Type *Ty) {
   return Run(Ty);
 }
 
+static bool isManifestConstant(const Constant *c) {
```

Comment:
This seems like a plain copy of an LLVM function.  This function is not exported by LLVM, so I'm wondering if it would be better to call a higher level function (something from `llvm/include/llvm/Analysis/ConstantFolding.h` perhaps)?  That will also give the benefit of actually folding anything it can.

Reply:
Looking at https://llvm.org/doxygen/ConstantFolding_8h.html I cannot find anything that can actually be used here. I mean we don't know what exactly is passed `Intrinsic::is_constant` and there is no generic function `tryToFoldWhateverItIs`

Reply:
Hi, @svenvh 
Yes, "isManifestConstant" is a plain copy of LLVM function from ConstantFolding.cpp. I understand that code duplicate won't make us happy here. However, "isManifestConstant" is a very simple function with few lines of code, it seems that bringing "isManifestConstant" into spirv translator won't introduce much code redundancy.
And I searched the invocations to "isManifestConstant" in llvm-project but couldn't find a general higher level function which meets our requirements here. "isManifestConstant" is only invoked by "ConstantFoldScalarCall1" which is not exported by llvm either.
Thank you very much.

Reply:
I agree it's a small function so from that perspective it shouldn't be much of a problem.  A remaining concern is on the attribution/licensing side: I am not a lawyer, but this file seems to carry an outdated header so I think we should be careful about what we add to this file.

Reply:
Hi, @svenvh 
Thanks very much for your reminder, do we need to go through some legal process for this patch?
Thanks very much.

Reply:
@svenvh, after discussion with @AlexeySachkov, I've factored out the "borrowed" code into a separate utility header. I've given that an Apache 2.0 w/ LLVM exceptions header (same as in LLVM's `lib/Analysis/ConstantFolding.cpp`). Could you please take a look at the new approach and approve the PR if this looks safe enough?

Reply:
@AGindinson thanks.  I am not a lawyer, but this looks like a reasonable solution.  Please make sure to squash this PR into one commit, so that the wrong license header does not end up in the git history.

Other than a minor comment posted below, I'll be happy to approve this PR.

Reply:
Applied both the comment and the squash. Thanks!